### ISSUE TITLE
Specific types not enforcing types

### ISSUE THREAD

**Issue by apense** [nim-lang/Nim#2814]
Sorry for the confusing title. What I mean is:

``` nim
echo ((high(int8) + 1).int8)
```

is producing `128`. If I'm not mistaken, and I haven't made a silly mistake (good chance I have), this should actually be throwing an error, as 128 is outside the range of `int8`. 
On the other hand, being explicit, like

``` nim
echo (128.int8)
```

_does_ cause a conversion error.

What's more, nothing seems to catch the error. That is,

``` nim
const i8: int8 = high(int8) + 1
echo i8 # produces 128
echo (i8 is int8) # produces true!!
echo high(int8) # checking myself, produces 127
```

Here's a range of tests:

``` nim
const 
  i8: int8 = high(int8) + 1
  i16: int16 = high(int16) + 1
  i32: int32 = high(int32) + 1
  i64: int64 = high(int64) + 1
echo i8 # 128
echo i16 # 32768
echo i32 # 2147483648
echo i64 # overflows to -9223372036854775808
```

Note that the `uints` do it too (`uint8` goes to 256). I can't test with `uint64`, since it doesn't have a `high` function.

For a related problem, see the following:

This is different from the original issue in that even being explicit doesn't raise an error if the type is involved in the expression.

``` nim
echo int8(int8(128) + int8(-129)) # gives "-1" without error
echo int8(128) # raises error
```

The compiler doesn't seem to care about the types inside the first expression, which can lead to some unexpected results.

``` nim
echo int8(int8(-128) + int8(-1)) # gives -129, which is outside low(int8) == -128
```

In short, specifying type only seems to work sometimes.

**Comment by apense** [comment]
Note that the `uint`s do it too (uint8 goes to 256). I can't test with `uint64`, since it doesn't have a `high` function.

**Comment by apense** [comment]
For a related problem, see the following:

```nim
echo int8(int8(128) + int8(-129)) # gives "-1" without error
echo int8(128) # raises error
```

The compiler doesn't seem to care about the types inside the first expression, which can lead to some unexpected results.

```nim
echo int8(int8(-128) + int8(-1)) # gives -129, which is outside low(int8) == -128
```

In short, specifying type only seems to work sometimes.

**Comment by Araq** [comment]
var x: int8 = high(int8)
echo(x + 1)

--> overflow. So it DOES work at runtime.
